### ISSUE TITLE
Provide enhanced DataSourceConfiguration support

### ISSUE THREAD

**Issue by btiernay** [spring-projects/spring-boot#1634]
Additional Configuration Properties
---

It would be nice to have more configuration properties available in `DataSourceConfiguration` similar to what is available in Dropwizard's DB module:

https://dropwizard.github.io/dropwizard/manual/jdbi.html#man-jdbi

```yaml
database:
  # the name of your JDBC driver
  driverClass: org.postgresql.Driver

  # the username
  user: pg-user

  # the password
  password: iAMs00perSecrEET

  # the JDBC URL
  url: jdbc:postgresql://db.example.com/db-prod

  # any properties specific to your JDBC driver:
  properties:
    charSet: UTF-8

  # the maximum amount of time to wait on an empty pool before throwing an exception
  maxWaitForConnection: 1s

  # the SQL query to run when validating a connection's liveness
  validationQuery: "/* MyService Health Check */ SELECT 1"

  # the minimum number of connections to keep open
  minSize: 8

  # the maximum number of connections to keep open
  maxSize: 32

  # whether or not idle connections should be validated
  checkConnectionWhileIdle: false

  # the amount of time to sleep between runs of the idle connection validation, abandoned cleaner and idle pool resizing
  evictionInterval: 10s

  # the minimum amount of time an connection must sit idle in the pool before it is eligible for eviction
  minIdleTime: 1 minute
```

Notice the ability to add JDBC properties, pool configuration, etc:

- https://github.com/dropwizard/dropwizard/blob/2655ac703a0f6ddb4eae933a74dae3d12fde287f/dropwizard-db/src/main/java/io/dropwizard/db/DataSourceFactory.java
- https://github.com/dropwizard/dropwizard/blob/2655ac703a0f6ddb4eae933a74dae3d12fde287f/dropwizard-db/src/main/java/io/dropwizard/db/DatabaseConfiguration.java

Consistent Configuration Prefix
---

I would also recommend moving the configuration point from `spring.datasource` to `datasource` to be consistent with `server` and `logging`. I see this as not being related to Spring proper as is the case with something like `spring.profiles.active`.

**Comment by wilkinsona** [comment]
DataSource configuration has been overhauled in 1.4 to make it clearer how to configure pool-specific properties. See https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-1.4.0-M1-Release-Notes#datasource-binding for details.
